qiskit/visualization/pulse_v2/generators.py

```diff
@@ -0,0 +1,683 @@
+by those generators. A stylesheet provides a list of generators and the core drawing
+function calls every generators for every instructions formatted in the input data format.
+
+An end-user can write arbitrary functions with the following function signature:
+
+    ```python
+    def my_object_generator(inst_data: InstructionTuple) -> List[drawing_objects.TextData]:
+        texts = []
+        # create some text data
+        text.append(TextData(data_type='custom', channel=inst_data.inst.channel, ...))
+
+        return texts
+    ```
+
+The user-defined drawing object is created by adding the generator to the stylesheet:
+
+    ```python
+    my_custom_style = {
+        'generator': {'waveform': [my_object_generator, ...]}
+    }
+    ```
+
+The user can set the custom stylesheet to the drawer interface.
+"""
+
+import re
+from collections import OrderedDict
+from fractions import Fraction
+from typing import Dict, Tuple, Any, List, Union
+
+import numpy as np
+
+from qiskit import pulse
+from qiskit.pulse import instructions
+from qiskit.visualization.exceptions import VisualizationError
+from qiskit.visualization.pulse_v2 import drawing_objects, data_types, PULSE_STYLE
+
+
+# Waveform related information generation
+
+
+def _parse_waveform(inst_data: data_types.InstructionTuple) \
+        -> Tuple[np.ndarray, np.ndarray, Dict[str, Any]]:
+    r"""A helper function that generates sample data array of the waveform with
+    instruction meta data.
+
+    Args:
+        inst_data: Instruction data set
+
+    Raises:
+        VisualizationError: When invalid instruction type is loaded.
+
+    Returns:
+        A tuple of xy data and metadata dictionary.
+    """
+    inst = inst_data.inst
+
+    meta = OrderedDict()
+    if isinstance(inst, instructions.Play):
+        # pulse
+        if isinstance(inst.pulse, pulse.ParametricPulse):
+            pulse_data = inst.pulse.get_sample_pulse()
+            meta.update(inst.pulse.parameters)
+        else:
+            pulse_data = inst.pulse
+        xdata = np.arange(pulse_data.duration) + inst_data.t0
+        ydata = pulse_data.samples
+    elif isinstance(inst, instructions.Delay):
+        # delay
+        xdata = np.arange(inst.duration) + inst_data.t0
+        ydata = np.zeros(inst.duration)
+    elif isinstance(inst, instructions.Acquire):
+        # acquire
+        xdata = np.arange(inst.duration) + inst_data.t0
+        ydata = np.ones(inst.duration)
+        acq_data = {'memory slot': inst.mem_slot.name,
+                    'register slot': inst.reg_slot.name if inst.reg_slot else 'N/A',
+                    'discriminator': inst.discriminator.name if inst.discriminator else 'N/A',
+                    'kernel': inst.kernel.name if inst.kernel else 'N/A'}
+        meta.update(acq_data)
+    else:
+        raise VisualizationError('Instruction %s cannot be drawn by filled envelope.' % type(inst))
+
+    meta.update({'duration (cycle time)': inst.duration,
```

Comment:
There is no standard format for metadata so that we can add arbitrary information, i.e. user may create a waveform object generator to add some analysis of given waveform as a part of metadata.